test/test_optim.py

```diff
@@ -1291,6 +1301,70 @@ def closure():
             res2 = optim_neg_inf.step(closure)
             self.assertEqual(type(res1), type(res2))
 
+    @onlyCPU
+    @skipIfTorchDynamo("https://github.com/pytorch/pytorch/issues/123238")
+    def test_grad_scaling_autocast_fused_optimizers(self):
+        # This ut is from test_cuda.py test_grad_scaling_autocast_fused_optimizers
+        # but only test Adam/AdamW on CPU
+        # TODO: haozhe, support SGD and unified this ut with the CUDA only one
+        optimizer_ctor = (torch.optim.Adam, torch.optim.AdamW)
+        optimizer_kwargs = ({"fused": True, "amsgrad": False}, {"fused": True, "amsgrad": True})
+        separate_unscale = (False, True)
+        for optim, kwargs, _separate_unscale in list(product(optimizer_ctor, optimizer_kwargs, separate_unscale)):
+            self._grad_scaling_autocast_fused_optimizers(
+                optimizer_ctor=optim, optimizer_kwargs=kwargs, separate_unscale=_separate_unscale)
+
+    def _grad_scaling_autocast_fused_optimizers(self, optimizer_ctor, optimizer_kwargs, separate_unscale):
+        (
+            mod_control, mod_scaling, opt_control, opt_scaling, data, loss_fn, _,
+        ) = _create_scaling_case(optimizer_ctor=optimizer_ctor, optimizer_kwargs=optimizer_kwargs, device='cpu')
+        kwargs = deepcopy(optimizer_kwargs)
+        kwargs["fused"] = False
+        opt_control = optimizer_ctor(mod_control.parameters(), lr=1.0, **kwargs)
+
+        scaler = torch.cpu.amp.GradScaler(init_scale=128.0)
+        tol = {'atol' : 1e-3, 'rtol' : 1e-3}
```

Comment:
This is very high...do we know why?

Reply:
Previously I have done some debug for `shapes=(1, 1)` and I observed some `diffs` brought by compiler optimization (as I described in PR description box.  So I think we may not able to make `fused` kernel exactly same with `non-fused` kernel and the multiple iters will accmulate the diffs.

But you are right that `1e-3` might too high. So I try more comparison between them and now they should be aligned
 (except compiler optimization for scalar case), what I further done is:
 - take more care on the precision between `double` and `float` conversion to exactly match non-fused kernel
```
// double lr, weight_decay, beta2
// float ptr for param_ptr, exp_avg_sq_ptr
// case1
param_ptr[d] = param_ptr[d] * (1 - float(lr) * float(weight_decay)); -> 
param_ptr[d] = param_ptr[d] * float(1 - lr * weight_decay);
// case2
exp_avg_sq_ptr[d] = exp_avg_sq_ptr[d] * beta2; ->
exp_avg_sq_ptr[d] = exp_avg_sq_ptr[d] * float(beta2);
```
 - For `lerp`, I copied the same code from [lerp_vec_kernel](https://github.com/pytorch/pytorch/blob/d04957c0c682d766987cad07dce20986ca4a5b78/aten/src/ATen/native/cpu/LerpKernel.cpp#L55-L61) instead of just finish the math from [document](https://github.com/pytorch/pytorch/blob/d04957c0c682d766987cad07dce20986ca4a5b78/aten/src/ATen/native/cpu/LerpKernel.cpp#L55-L61)

Reply:
Yea, it's reasonable the fused will not be the same as the non-fused due to intermediates building up differences. Thanks for the investigation here and aligning them!